core/base/topologicalOptimization/TopologicalOptimization.h

```diff
@@ -0,0 +1,1674 @@
+      birthPairToDeleteTargetDiagram.push_back(
+        (pair.birth.sfValue + pair.death.sfValue) / 2);
+      deathPairToDeleteCurrentDiagram.push_back(
+        static_cast<SimplexId>(pair.death.id));
+      deathPairToDeleteTargetDiagram.push_back(
+        (pair.birth.sfValue + pair.death.sfValue) / 2);
+    }
+
+    for(const auto &entry : matchingsBlockPairs) {
+      // Delete pairs that have no equivalence
+      if(entry.size() == 1) {
+        birthPairToDeleteCurrentDiagram.push_back(
+          static_cast<SimplexId>(entry[0].birth.id));
+        birthPairToDeleteTargetDiagram.push_back(
+          (entry[0].birth.sfValue + entry[0].death.sfValue) / 2);
+        deathPairToDeleteCurrentDiagram.push_back(
+          static_cast<SimplexId>(entry[0].death.id));
+        deathPairToDeleteTargetDiagram.push_back(
+          (entry[0].birth.sfValue + entry[0].death.sfValue) / 2);
+        continue;
+      } else if(entry.empty())
+        continue;
+
+      SimplexId valueBirthPairToChangeCurrentDiagram
+        = static_cast<SimplexId>(entry[0].birth.id);
+      SimplexId valueDeathPairToChangeCurrentDiagram
+        = static_cast<SimplexId>(entry[0].death.id);
+
+      double valueBirthPairToChangeTargetDiagram = entry[1].birth.sfValue;
+      double valueDeathPairToChangeTargetDiagram = entry[1].death.sfValue;
+
+      birthPairToChangeCurrentDiagram.push_back(
+        valueBirthPairToChangeCurrentDiagram);
+      birthPairToChangeTargetDiagram.push_back(
+        valueBirthPairToChangeTargetDiagram);
+      deathPairToChangeCurrentDiagram.push_back(
+        valueDeathPairToChangeCurrentDiagram);
+      deathPairToChangeTargetDiagram.push_back(
+        valueDeathPairToChangeTargetDiagram);
+    }
+  }
+}
+
+/*
+  This function allows you to copy the values of a pytorch tensor
+  to a vector in an optimized way.
+*/
+#ifdef TTK_ENABLE_TORCH
+int ttk::TopologicalOptimization::tensorToVectorFast(
+  const torch::Tensor &tensor, std::vector<double> &result) const {
+  TORCH_CHECK(
+    tensor.dtype() == torch::kDouble, "The tensor must be of double type");
+  const double *dataPtr = tensor.data_ptr<double>();
+  result.assign(dataPtr, dataPtr + tensor.numel());
+
+  return 0;
+}
+#endif
+
+template <typename dataType, typename triangulationType>
+int ttk::TopologicalOptimization::execute(
+  const dataType *const inputScalars,
+  dataType *const outputScalars,
+  SimplexId *const inputOffsets,
+  triangulationType *triangulation,
+  const ttk::DiagramType &constraintDiagram) const {
+
+  Timer t;
+  double stoppingCondition = 0;
+  bool enableTorch = true;
+
+  if(methodOptimization_ == 1) {
+#ifndef TTK_ENABLE_TORCH
+    this->printWrn("Adam unavailable (Torch not found).");
+    this->printWrn("Using direct gradient descent.");
+    enableTorch = false;
+#endif
+  }
+
+  //=======================
+  //    Copy input data
+  //=======================
+  std::vector<double> dataVector(vertexNumber_);
```

Comment:
Why is the data temporarily converted to double for the computation?
Is that a requirement from Torch?